roms_tools/setup/cdr_forcing.py

```diff
@@ -21,171 +18,210 @@
-                raise ValueError(
-                    f"First time in 'times' cannot be before start_time ({self.start_time})."
-                )
+        if times[-1] > self.end_time:
+            raise ValueError(
+                f"Last time in release '{self.release.name}' is after end_time ({self.end_time})."
+            )
 
-            # Last time must not be after end_time
-            if self.times[-1] > self.end_time:
-                raise ValueError(
-                    f"Last time in 'times' cannot be after end_time ({self.end_time})."
-                )
         return self
 
     @model_validator(mode="after")
-    def check_volume_fluxes(self) -> "Release":
-        """Ensure 'volume_fluxes' matches length of 'times'."""
+    def validate_release_location(self) -> "ReleaseSimulationManager":
+        """Ensure the release is consistent with the simulation grid."""
+        _validate_release_location(self.grid, self.release)
+        return self
 
-        num_times = len(self.times)
+    @model_validator(mode="after")
+    def extend_to_endpoints(self) -> "ReleaseSimulationManager":
+        """Extend the release time series to include the simulation time endpoints."""
+        self.release._extend_to_endpoints(self.start_time, self.end_time)
+        return self
 
-        if isinstance(self.volume_fluxes, list):
-            if len(self.volume_fluxes) != num_times:
-                raise ValueError(
-                    f"The length of 'volume_fluxes' ({len(self.volume_fluxes)}) does not match the number of times ({num_times})."
-                )
 
-        return self
+class ReleaseCollector(BaseModel):
+    """Collects and validates multiple releases against each other."""
+
+    releases: List[Release]
 
     @model_validator(mode="after")
-    def check_tracer_concentrations(self) -> "Release":
-        """Ensure tracer concentractions match length of 'times'."""
+    def check_unique_name(self) -> "ReleaseCollector":
+        """Check that all releases have unique names."""
+        names = [release.name for release in self.releases]
+        duplicates = [name for name, count in Counter(names).items() if count > 1]
 
-        num_times = len(self.times)
+        if duplicates:
+            raise ValueError(
+                f"Multiple releases share the following name(s): {', '.join(repr(d) for d in duplicates)}. "
+                "Each release must have a unique name."
+            )
 
-        for tracer, val in self.tracer_concentrations.items():
+        return self
 
-            if isinstance(val, list):
-                if len(val) != num_times:
-                    raise ValueError(
-                        f"The length of tracer '{tracer}' ({len(val)}) does not match the number of times ({num_times})."
-                    )
+    def determine_release_type(self) -> "ReleaseCollector":
+        """Ensure all releases are of the same type, and set the release_type."""
+        if not all(
```

Comment:
consider a set trick to make it clearer that we're interested in membership size:

this helps to avoid asking "wait, why am i looking at the 0-th item?" when you read this later on, and it lets you log the underlying issue (which someone will inevitably want to fix the bug). 

It also gives you something you can return without asking for type info again.


```
release_types = set(map(type, self.releases))
if len(release_types) > 1:
    type_list = ", ".join(release_types)
    raise ValueError(f"Not all releases have the same type. Received: {type_list}")

return release_types.pop()
```

Reply:
Good point and great suggestion. Thanks! Changed in [49ef5cb](https://github.com/CWorthy-ocean/roms-tools/pull/301/commits/49ef5cbacacd4311e673856c7c044dbfe4c7e68b).